Basic Linked list implementation in c++

SeanSweeney04 opened the pull request.

gregory-sokolov commented on LinkedList/LinkedList.h:
```diff
@@ -0,0 +1,25 @@
+#ifndef LINKEDLIST_H
+#define LINKEDLIST_H
+#include <iostream>
```
Library file includes are usually not placed into your header files in order not to introduce extra dependencies. LinkedList is a data structure, it's not supposed to use io-streams that's why no reason to include it here.

SeanSweeney04 replied:
Removed library file from LinkedList.h

SeanSweeney04 replied:
removed

gregory-sokolov commented on LinkedList/LinkedList.h:
```diff
@@ -0,0 +1,25 @@
+#ifndef LINKEDLIST_H
+#define LINKEDLIST_H
```
It's better to place an empty line after include guards.

SeanSweeney04 replied:
![Screenshot 2025-03-13 101624](https://github.com/user-attachments/assets/f95e1ff0-243d-4f9d-a05b-8fc769dc5386)

SeanSweeney04 replied:
added empty line after include guards

gregory-sokolov commented on LinkedList/LinkedList.h:
```diff
@@ -0,0 +1,25 @@
+
+#include "Node.h"
+
+class LinkedList {
```
Do not use Egyptian style. We use:
class A
{
    ...
}

SeanSweeney04 replied:
Changed to correct format

SeanSweeney04 replied:
Fixed

gregory-sokolov commented on LinkedList/LinkedList.h:
```diff
@@ -0,0 +1,25 @@
+
+class LinkedList {
+
+private:
```
Class layout must be formed "from general to particular" - public first, private last.

SeanSweeney04 replied:
![Screenshot 2025-03-13 101745](https://github.com/user-attachments/assets/52014dd2-8013-4e3f-a472-30a5240308df)

SeanSweeney04 replied:
Changed to correct order

gregory-sokolov commented on LinkedList/LinkedList.h:
```diff
@@ -0,0 +1,25 @@
+class LinkedList {
+
+private:
+    Node* head = NULL;
```
In modern C++, use nullptr instead of NULL.

SeanSweeney04 replied:
![Screenshot 2025-03-13 101745](https://github.com/user-attachments/assets/39255e73-d6e9-4224-a697-18c9ab5bfa56)

SeanSweeney04 replied:
Changed NULL to nullptr

gregory-sokolov commented on LinkedList/LinkedList.h:
```diff
@@ -0,0 +1,25 @@
+
+public:
+    LinkedList();
+    LinkedList(const LinkedList& list);
```
Parametrized ctor is missing, operator= is missing.

SeanSweeney04 replied:
added both missing constructors

gregory-sokolov commented on LinkedList/LinkedList.h:
```diff
@@ -0,0 +1,25 @@
+    void remove(int x);
+    void print();
+   
+    ~LinkedList();
```
Destructor must be placed right after all the constructors.

SeanSweeney04 replied:
Moved destructor to after constructors

gregory-sokolov commented on LinkedList/LinkedList.h:
```diff
@@ -0,0 +1,25 @@
+  
+    void insert(int x);
+    void remove(int x);
+    void print();
```
The functionality of displaying the LinkedList with print() must not be a part of the data structure.
This is a typical class design fault: to mix UI and business logic. I understand that it is easier to do it in the member function, but it is definitely not the LinkedList responsibility to handle UI interactions. Its resposibility is to handle it's internal data. Displaying the data is the responsibility of some other class/module, e.g. LinkedListLogger.

SeanSweeney04 replied:
Added an additional LinkedListLogger class to handle this

gregory-sokolov commented on LinkedList/Node.h:
```diff
@@ -0,0 +1,12 @@
+#ifndef NODE_H
+#define NODE_H
+
+class Node {
```
Making fields public in a class breaks incapsulation. Remember what was the difference between the class and the struct.

gregory-sokolov commented on LinkedList/Node.h:
```diff
@@ -0,0 +1,12 @@
+public:
+    int data;
+    Node* next;
+    Node(int x);
```
The same comment about the class layout "from general to particular" - constructors go first, fields go last. It's better to separate constructor block, function block and data fields with empty line.

SeanSweeney04 replied:
Changed to correct order

gregory-sokolov commented on LinkedList/Logger.h:
```diff
@@ -0,0 +1,21 @@
+#pragma once
```
Either pragma or include guards but neither nor both.

SeanSweeney04 replied:
removed pragma

gregory-sokolov commented on LinkedList/Logger.h:
```diff
@@ -0,0 +1,21 @@
+    static void enableFileLogging(const std::string& filename);
+
+private:
+    static std::ofstream logFile;
```
static is usually a 'red flag' that signals about class design mistakes. Can you rewrite this class and its usages in main without using static?

SeanSweeney04 replied:
Changed Logger from a static utility to an instance-based class.

Removed static methods and instead passed a Logger& instance to LinkedList.
This improves encapsulation, allows multiple loggers, and makes testing easier.
Updated main.cpp to pass a Logger instance to LinkedList.
Replaced Logger::log(...) calls with logger.log(...) in LinkedList.cpp.

gregory-sokolov commented on LinkedList/LinkedList.cpp:
```diff
@@ -0,0 +1,134 @@
+// LinkedList.cpp : This file contains the 'main' function. Program execution begins and ends there.
```
Remove all auto-generated comments from Visual Studio in the file.

gregory-sokolov commented on LinkedList/LinkedList.cpp:
```diff
@@ -0,0 +1,134 @@
+#include "LinkedList.h"
+#include "Node.h"
+#include "Logger.h"
+using namespace std;
```
"using" must be separated from includes with an empty line.

SeanSweeney04 replied:
Added empty line after using

gregory-sokolov commented on LinkedList/LinkedList.cpp:
```diff
@@ -0,0 +1,134 @@
+using namespace std;
+
+// Node class constructor: Initializes a node with given data and sets next to NULL
+
```
No need to put an empty line after the function comment.

SeanSweeney04 replied:
Removed empty line

gregory-sokolov commented on LinkedList/LinkedList.cpp:
```diff
@@ -0,0 +1,134 @@
+}
+
+// Default constructor for LinkedList: Initializes head and tail as NULL
+LinkedList:: LinkedList() { 
```
Egyptian style.

SeanSweeney04 replied:
Changed to correct format

gregory-sokolov commented on LinkedList/LinkedList.cpp:
```diff
@@ -0,0 +1,134 @@
+// Copy constructor: Creates a deep copy of the given LinkedList
+LinkedList::LinkedList(const LinkedList& list)
+{
+    Logger::log(LogLevel::INFO, "Copy Constructor called.");
```
I would recommend to insert empty lines after logical code blocks within the function. For example, you placed some comments below, it looks like each comment start a new logical group of code lines that should be separated with empty lines. It improves readability.

SeanSweeney04 replied:
So like this? 
// Node class constructor: Initializes a node with given data and sets next to nullptr

Node::Node(int x)

gregory-sokolov commented on LinkedList/LinkedList.cpp:
```diff
@@ -0,0 +1,134 @@
+Node::Node(int x)
+{
+    data = x;
+    next = NULL;
```
NULL -> nullptr, please change it everywhere.

SeanSweeney04 replied:
Changed all NULL to nullptr

gregory-sokolov commented on LinkedList/LinkedList.cpp:
```diff
@@ -0,0 +1,134 @@
+            head = newNode;
+            tail = newNode;
+        }
+        else { // Append the new node at the end
```
Egyptian style. Should be:
if
{
}
else
{
}

SeanSweeney04 replied:
Fixed

gregory-sokolov commented on LinkedList/LinkedList.cpp:
```diff
@@ -0,0 +1,134 @@
+        if (cur->data == x) {
+            Logger::log(LogLevel::INFO, "Removing: " + std::to_string(x));
+            prev->next = cur->next; // Bypass the node
+            if (cur == tail) tail = prev; // Update tail if needed
```
Even if there is only one line in if-else-for-while body, it must be placed in braces.
if
{
...
}

SeanSweeney04 replied:
added curly brackets to if statement

gregory-sokolov commented on LinkedList/LinkedList.cpp:
```diff
@@ -0,0 +1,134 @@
+        cur = cur->next;
+    }
+    Logger::log(LogLevel::WARNING, "Element not found: " + std::to_string(x));
+
```
Extra empty line.

SeanSweeney04 replied:
Removed

gregory-sokolov commented on LinkedList/LinkedList.cpp:
```diff
@@ -0,0 +1,134 @@
+        temp = temp->next;
+    }
+    cout << endl;
+
```
Extra empty line before }.

SeanSweeney04 replied:
removed

gregory-sokolov commented on LinkedList/LinkedList.cpp:
```diff
@@ -0,0 +1,134 @@
+// Function to print all nodes in the linked list
+void LinkedList::print()
+{
+    Logger::log(LogLevel::INFO, "Destructor called.");
```
Copy-paste error? print is not destructor.

SeanSweeney04 replied:
My bad, moved logger to destructor

gregory-sokolov commented on LinkedList/Logger.cpp:
```diff
@@ -0,0 +1,38 @@
+#include "Logger.h"
```
Good design decision to introduce a Logger class!

gregory-sokolov commented on LinkedList/Logger.h:
```diff
@@ -0,0 +1,21 @@
+#include <fstream>
+#include <string>
+
+enum class LogLevel { INFO, WARNING, ERROR };
```
Enums are declared without 'class' keyword.

SeanSweeney04 replied:
removed class keyword

gregory-sokolov commented on LinkedList/main.cpp:
```diff
@@ -0,0 +1,74 @@
+    l3->print();
+    delete l3; // Destructor should be called
+
+    return 0;
```
You tested the LinkedList functionality, and now you need to implements copy ctor and copy assignment and add some tests with them. It's interesting to explore some cases with copying objects, passing/returning them from fuction, going in-out of scope, etc.

SeanSweeney04 replied:
Updated main.cpp to support instance-based logging.

Now explicitly passes a Logger& instance to LinkedList at creation.
Ensures all LinkedList objects receive a Logger reference, fixing constructor errors.
Maintains proper encapsulation while allowing flexibility in logging.

gregory-sokolov commented:
General comment. Good implementation for the data structure itself, really nice idea of using separate logger class for logging purposes. However, displaying the contents of the LinkedList must not be the part of the class, because it's not the list responsibility to deal with UI (console output). It should be done in somewhere outside, in main or LinkedList logger.
Also, in C++ we don't use egyptian style, so please correct it everywhere.

SeanSweeney04 commented:
Refactored Logger to be instance-based (removed static methods).
Updated main.cpp to correctly pass a Logger& instance to LinkedList.
Separated print() from LinkedList to follow Single Responsibility Principle:

Created a new LinkedListLogger class to handle printing.
LinkedListLogger now accesses LinkedList via a friend class declaration.
Encapsulated Node class fields to enforce data protection.
Improved project structure by adding new files and updating includes.

Commit:
Reinitialized Git and restored project
